ENH: stats: multivariate Wasserstein distance/EMD

com3dian opened the pull request:
#### Reference issue
Closes gh-17290

#### What does this implement/fix?
Extend the current `stats.wasserstein_distance` with linear programming method `optimize.milp`.
The Linear programming implementation is only used under n-D case as the `stats._cdf_distance` is faster on 1-D data.

#### Additional information
Due to the limit of `milp`, the solution can only be found when the optimal transport plan has less than 10^9 variables.

com3dian committed:
Feature: Enhancement: issue 17290
Expand the stats/_stats_py.wasserstein_distance to multi-dimensional by
applying linear programmming method(optimize.milp) and keeping the
signature not changed.
In 1_d case the function still calls _stats_py._cdf_distance as it is
faster than linear programming.

com3dian committed:
Feature: ENH: (#17290)

* Add n-D tests under stats/tests/tests_stats.TestWassersteinDistance class

* Add method test_multi_dim under TestWassersteinDistance class

com3dian committed:
Feature: ENH: (#17290)

* Add test_rotation_reflection method under stats/tests/test_stats.TestWassersteinDistance class

mdhaber commented:
Seeing what failed often requires several clicks:

![image](https://user-images.githubusercontent.com/6570539/203838679-6092b41c-27aa-408e-9e54-f5ddeb10381f.png)

![image](https://user-images.githubusercontent.com/6570539/203838740-ba67f513-0687-46a5-857e-40efea9d63f5.png)

![image](https://user-images.githubusercontent.com/6570539/203838788-8ca9c693-f744-415d-9a9d-edd0d8f8fcca.png)

![image](https://user-images.githubusercontent.com/6570539/203838865-26fa7e1e-be39-494a-a320-5415428ed949.png)

```
=================================== FAILURES ===================================
________________ TestWassersteinDistance.test_same_distribution ________________
[gw1] linux -- Python 3.9.15 /opt/hostedtoolcache/Python/3.9.15/x64/bin/python
/opt/hostedtoolcache/Python/3.9.15/x64/lib/python3.9/site-packages/scipy/stats/tests/test_stats.py:7212: in test_same_distribution
    assert_equal(stats.wasserstein_distance(
E   AssertionError: 
E   Items are not equal:
E    ACTUAL: 1.0839226509442299e-17
E    DESIRED: 0
        self       = <scipy.stats.tests.test_stats.TestWassersteinDistance object at 0x7f1b8394dd00>
```

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -7201,10 +7207,14 @@ def test_same_distribution(self):
             0)
+        assert_equal(stats.wasserstein_distance(
+            [[0, 0], [15, 0.2]], [[15, 0.2], [0, 0]]), 0)
+        assert_equal(stats.wasserstein_distance(
```
Here's the failure. Of course, they are essentially equal, with an error of `1.0839226509442299e-17`. If it's not something that can be fixed, use `assert_allclose(result, 0, atol=1e-16)`.

com3dian replied:
Sure I would look at the lint tools, but I assume it's also okay if i manually change all the lint failures?

mdhaber replied:
You _should_ manually fix the failures. The lint tools will just help you check whether you actually fixed them all.

com3dian committed:
Merge remote-tracking branch 'refs/remotes/upstream/main' into ENH_17290

com3dian committed:
BUG: PR_17473

* Fixed typo and linting

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9042,14 +9051,58 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
+    called the optimal transport problem or the Monge problem, which can be
+    expressed in this form,
+
+    Given :math:`\Gamma` the transport plan, :math:`D` the distance matrix and
```
This doesn't seem to be rendering:

![image](https://user-images.githubusercontent.com/6570539/204123357-18337122-b40e-41fe-8cb2-7fe6bbdd929a.png)

You can see the rendered docs by clicking:
![image](https://user-images.githubusercontent.com/6570539/204123378-d25569c2-50dd-47df-8a81-d75562b0de12.png)

mdhaber replied:
Better!

com3dian replied:
Thanks Matt :rofl: !

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9042,14 +9051,58 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
+    Given :math:`\Gamma` the transport plan, :math:`D` the distance matrix and
+    follows,
+
+    .. math:
```
```suggestion
    .. math::
```
I think that's the problem.

com3dian replied:
Thank you! I will try to fix all and come back to you next week.

mdhaber reviewed (commented):
Here's a first pass. One comment I forgot to add inline - it would be good to include at least one complex problem that can be checked against POT. 

Next time I'll take a closer look at the implementation.

com3dian commented on scipy/stats/_stats_py.py:
```diff
@@ -9002,19 +9005,25 @@ def combine_pvalues(pvalues, method='fisher', weights=None):
     ----------
     u_values, v_values : array_like
-        Values observed in the (empirical) distribution.
+        could be:
+        - Values observed in the (empirical) distribution.
+        - Value of each class in the (distribution)histogram.
```
You are right with most of your points. It is true that 2D input should be like [vec1, vec2, ...] with each vector has same length(dimensions). And I do automatically convert any 1D input into 2D so that the dimension of both input can be easily compared. 
You are totally right about the uniform distribution example, in the context of statistics, the input values are 'outcomes' and weights are 'frequencies', in machine learning context in they can also be understood as 'data points' and 'weights'. I think I can add this in the document and also a non-unique example, if it what you mean here.

mdhaber replied:
I was hoping you could elaborate in the documentation, then. Otherwise, users will have the same questions.

com3dian commented on scipy/stats/tests/test_stats.py:
```diff
@@ -7201,10 +7207,18 @@ def test_same_distribution(self):
+            [1, 1, 1, 1, 1], [1, 2, 2]), 0)
 
     def test_shift(self):
         # If the whole distribution is shifted by x, then the Wasserstein
```
Regarding all testing comments, do you think it's better to replace all old tests with random data or only design random tests for our new feature?

mdhaber replied:
Only the new features is fine. If you'd like to strengthen the existing tests in a separate PR, that would be great.

com3dian replied:
good, I'm leaving old tests remain for now then.

com3dian commented on scipy/stats/tests/test_stats.py:
```diff
@@ -7192,6 +7192,12 @@ def test_simple(self):
+            1.8)
 
     def test_same_distribution(self):
         # Any distribution moved to itself should have a Wasserstein distance of
```
I was just following existing testing styles, it also making more sense to me if using random generated data to test. And I will be happy if you can show me an example for `pytest.mark.parametrize`, I don't quite get the idea of using this decorator performing tests. Thank you.

mdhaber replied:
Here's an example that's pretty involved but also pretty readable:
https://github.com/scipy/scipy/blob/c365b78109267dad1a2109f6aec51de35f42809d/scipy/stats/tests/test_resampling.py#L791-L809

Here, the p-value returned by `monte_carlo_test` is checked against all combinations of two hypothesis test functions, three alternative hypothesis settings, and several values of data skewness. The first argument of `paramatrize` is the name `pytest` should use for the keyword argument, and the second argument is a sequence of values you should pass. Stacking the decorators makes it perform all combinations.

Sometimes you don't want to perform all combinations. There are other ways of doing this, but I would usually do something like:
```python3
@pytest.mark.parametrize('args', [(stats.skewtest, 'less', -1)
                                  (stats.kurtosistest, 'greater', 2)])
 def test_against_normality_tests(self, args):
    hypotest, alternative, a = args
```

com3dian commented on scipy/stats/tests/test_stats.py:
```diff
@@ -7192,6 +7192,12 @@ def test_simple(self):
         assert_almost_equal(stats.wasserstein_distance(
             [0, 1, 2], [1, 2, 3]),
             1)
+        assert_almost_equal(stats.wasserstein_distance(
+            [[0, 1], [1, 0]], [[1, 1]], [1, 1], [1]),
+            1)
```
Okay I see what you mean here, well the next example is
```python
    assert_almost_equal(stats.wasserstein_distance(
        [[0, 0], [2, 1]], [[2, 0]], [4, 1], [1]),
        1.8)
```
and when the target distribution has only one value, the Wasserstein distance degenerates to the weighted sum of pair wise distance between each input value and the only value in the output. So there is 
```math
\begin{align*}
wass([[0, 0], [2, 1]], [[2, 0]], [4, 1], [1]) &= weightedsum([[0, 0], [2, 1]], [[2, 0]], weight = [[4, 1], [1]]) \\ 
&= \frac{4}{4+1} D([0, 0], [2, 0]) + \frac{1}{4+1} D([2, 1], [2, 0]) \\
&= \frac{4}{4+1} \times 2 + \frac{1}{4+1} \times 1 \\
&= 0.8 \times 2 + 0.2 \times 1 \\
&= 1.8
\end{align*}
```

mdhaber replied:
OK; please add that sort of thing as comments to the code. A reference for "the Wasserstein distance degenerates to the weighted sum of pair wise distance " - just a URL as a comment - would be great.

mdhaber replied:
This comment is still open.

com3dian commented on scipy/stats/tests/test_stats.py:
```diff
@@ -7264,6 +7325,49 @@ def test_inf_values(self):
+            stats.wasserstein_distance([[1, 1], [2, 1]], [[np.inf, -np.inf]],
+                                       [1, 1]),
+            np.inf)
+        with suppress_warnings() as sup:
+            sup.record(RuntimeWarning, "invalid value*")
```
Okay I can change this to like
```suggestion
        with np.errstate(invalid='ignore'):
            ...
```
but as I followed the existing tests format should I also change line 7324-7327? #

mdhaber replied:
Do this with everything new, but don't worry about changing any existing tests in this PR.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9059,8 +9112,14 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
     .. [1] "Wasserstein metric", https://en.wikipedia.org/wiki/Wasserstein_metric
-    .. [2] Ramdas, Garcia, Cuturi "On Wasserstein Two Sample Testing and Related
-           Families of Nonparametric Tests" (2015). :arXiv:`1509.02237`.
+    .. [2] https://lilianweng.github.io/posts/2017-08-20-gan/#what-is-
+           wasserstein-distance.
```
Here's an example of references for [another function](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.bootstrap.html):
![image](https://user-images.githubusercontent.com/6570539/210727080-ba1f4933-f49b-4abe-b21b-2d2cdeeb2077.png)

If you can add the author, I think this would be sufficient.

mdhaber replied:
This was marked as resolved, but it has not been resolved. The order is typically Author, name of article, name of publication, then URL.  So I would add the name of the author at the beginning of what you suggested.

com3dian replied:
Okay now it looks like
```
    .. [2] Lili Weng, "What is Wasserstein distance?", Lil'log, 
           https://lilianweng.github.io/posts/2017-08-20-gan/#what-is-
           wasserstein-distance.
```

mdhaber commented:
@com3dian just a reminder that you have a review waiting. If you are no longer able to finish this up, that's fine. If I don't hear back in about a week, I'll go ahead and close.

com3dian commented:
Hey Matt! Apologies for not being active on this during last months, I have updated some of the unit test using `pytest`, I will push my changes later this week so we can maybe discuss if it meet `scipy` requirements.

com3dian committed:
Merge branch 'scipy:main' into feature_emd_gh17290

com3dian committed:
* ENH-17290

* Add tests for nD shifting/rotation/reflection in TestWassersteinDistance class

* updated comment of the stats. wasserstein_distanc function

com3dian commented:
Hi Matt I added two methods(`test_shift_nD` and `test_rotation_reflection_nD`) in `TestWassersteinDistance` using `pytest` and changed the old fashion nD tests to something like this
```{python}
uv, vv = [[0, 0], [15, 0.2]], [[15, 0.2], [0, 0]]
        distance = stats.wasserstein_distance(uv, vv)
        assert_almost_equal(distance, 0)
```
However, some tests are throwing submodule errors
```
fatal: remote error: upload-pack: not our ref 2f83e9aceceba81a5e7c532ca2e9dcf697999db6
Errors during submodule fetch:
	scipy/_lib/boost
```
do you have any idea what is the cause and any walkaround?

com3dian committed:
ENH-17473

* fix rendering issue of stats.wasserstein_distance doc

com3dian committed:
Merge branch 'main' of https://github.com/scipy/scipy into feature_emd_gh17290

melissawm commented on scipy/stats/_stats_py.py:
```diff
@@ -9740,9 +9805,58 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
+
+    # get cost matrix
+    D = distance_matrix(u_values, v_values, p=2)
+    cost = D.reshape((n * m))
```
The linter is flagging this as "extraneous parenthesis"

com3dian replied:
Hi melissa, yes the linter is correct on this, single brackets `( n * m )` works fine with `reshape` and is equivalent with the current double brackets way. I'm doing this simply because Matt suggested [here](https://github.com/scipy/scipy/pull/17473#discussion_r1132655107).

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -7249,6 +7294,88 @@ def test_inf_values(self):
+        shift_result = stats.wasserstein_distance(uu_shift, vv_shift)
+        assert_almost_equal(expected, shift_result)
+
+    @pytest.mark.parametrize('input_dim', (3, 4, 5, 22, 40))
```
```suggestion
    @pytest.mark.parametrize('input_dim', (3, 4, 5, 22))
```
I think this is causing most of the CI failures on Azure Pipelines.
```
FAILED ..\..\stats\tests\test_stats.py::TestWassersteinDistance::test_rotation_reflection_nD[data_volumn0-40]
FAILED ..\..\stats\tests\test_stats.py::TestWassersteinDistance::test_rotation_reflection_nD[data_volumn1-40]
FAILED ..\..\stats\tests\test_stats.py::TestWassersteinDistance::test_rotation_reflection_nD[data_volumn3-40]
FAILED ..\..\stats\tests\test_stats.py::TestWassersteinDistance::test_rotation_reflection_nD[data_volumn2-40]
FAILED ..\..\stats\tests\test_stats.py::TestWassersteinDistance::test_rotation_reflection_nD[data_volumn4-40]
```
Specifically,
```
File "C:\hostedtoolcache\windows\Python\3.9.13\x86\lib\site-packages\scipy\stats\tests\test_stats.py", line 7373, in test_rotation_reflection_nD
    rotation_matrix = np.linalg.multi_dot(rotations)
  File "<__array_function__ internals>", line 180, in multi_dot
  File "C:\hostedtoolcache\windows\Python\3.9.13\x86\lib\site-packages\numpy\linalg\linalg.py", line 2726, in multi_dot
    order = _multi_dot_matrix_chain_order(arrays)
  File "C:\hostedtoolcache\windows\Python\3.9.13\x86\lib\site-packages\numpy\linalg\linalg.py", line 2792, in _multi_dot_matrix_chain_order
    q = m[i, k] + m[k+1, j] + p[i]*p[k+1]*p[j+1]

+++++++++++++++++++++++++++++++++++ Timeout ++++++++++++++++++++++++++++++++++++
.....[gw5] node down: Not properly terminated

```
It's probably just that the AZP resources are insufficient. (Do you expect that to be a demanding problem for `np.linalg.multi_dot`? If not, it might be worth filing a separate bug report.) In any case, it does not indicate a problem with `wasserstein_distance`, so let's just avoid it.

com3dian replied:
> Do you expect that to be a demanding problem for `np.linalg.multi_dot`
No, I don't foresee that. I will remove that in the next update.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9740,9 +9805,58 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
+
+    # get cost matrix
+    D = distance_matrix(u_values, v_values, p=2)
+    cost = D.reshape((n * m))
```
```suggestion
    cost = D.reshape(n * m)
```

mdhaber reviewed (commented):
Oops, I didn't see your ping. 

![image](https://user-images.githubusercontent.com/6570539/224383930-c7b55d0d-2cc6-4ba5-b15e-d1c9c7f96cd0.png)

Please do not force-push here; merge and regular-push instead. We will likely squash-merge this, so the history will be clean. Even if there are commits we want to preserve, we can always rebase at the end, once review is complete, rather than changing the history mid-review.

![image](https://user-images.githubusercontent.com/6570539/224387651-8336ab92-1389-4871-92a5-b460c97bc1a6.png)

Also, please leave comments I've made unresolved. I will mark them as resolved when I have reviewed the response or changes you've made.
 
Re:
```
fatal: remote error: upload-pack: not our ref 2f83e9aceceba81a5e7c532ca2e9dcf697999db6
Errors during submodule fetch:
	scipy/_lib/boost
```
you mean this is happening locally, right? Is it still happening after you merged main? Make sure you merge the latest version of `main`. The `boost` submodule has been replaced by `boost_math`. If you have trouble with that, let me know, and I can suggest how to fix it. (In short, navigate into `boost_math`, add [the boost math repo](https://github.com/boostorg/math) as a remote, and fetch it, then navigate back to the SciPy root directory and `git submodule update --init` before building.)

mdhaber committed:
Apply suggestions from code review

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9710,14 +9719,62 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
+
+    .. math::
+
+        x = vec(\Gamma)
+        c = vec(D)
+        b = \begin{bmatrix}
+                u\\
+                v\\
+            \end{bmatrix}
```
This renders like:
![image](https://user-images.githubusercontent.com/6570539/224386598-1c47ab35-2256-40c8-a40f-745540139d9f.png)

com3dian replied:
Thanks I will add line breaks

mdhaber replied:
That helped, but what does $vec$ mean? Currently, it renders as though it were the product of variables $v$, $e$, and $c$. If it is a function, it should not be italicized (e.g. $\sin x, \log x$), and I'd suggest defining what $\mbox{vec} \left( \cdot \right)$ means.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9710,14 +9719,62 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
+    solution :math:`y^*`), the Wasserstein distance :math:`l_1 (u, v)` can
+    be computed as :math:`b^T x^*`.
+
+    The above solution is inspired by Vincent Herrmann's blog [5].For a
```
```suggestion
    The above solution is inspired by Vincent Herrmann's blog [5]_. For a
```

com3dian replied:
Thanks!

mdhaber replied:
Sure thing. Please let me know when you think you have addressed all comments and this is ready for another look.

com3dian replied:
Thanks a lot! I think now it should be all clean and you can have look at the comments now, and if you have extra time could you also check the `TestWassersteinDistance` class (mostly the last two methods)? I have updated those two tests with `pytest` but don't know if it meets your standards :smile: .

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9002,19 +9005,25 @@ def combine_pvalues(pvalues, method='fisher', weights=None):
 def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
     r"""
-    Compute the first Wasserstein distance between two 1D distributions.
+    Compute the general wasserstein-1 function between two n dimensional
```
This was marked as resolved, but it has not been resolved.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9002,19 +9005,25 @@ def combine_pvalues(pvalues, method='fisher', weights=None):
-    seen as the minimum amount of "work" required to transform :math:`u` into
-    :math:`v`, where "work" is measured as the amount of distribution weight
-    that must be moved, multiplied by the distance it has to be moved.
+    The wasserstein distance, also called the Earth mover distance or the
```
This was marked as resolved, but Wasserstein still needs to be capitalized. If you disagree, please say so, and I will reply with evidence that it should be capitalized.

com3dian commented on scipy/stats/_stats_py.py:
```diff
@@ -9002,19 +9005,25 @@ def combine_pvalues(pvalues, method='fisher', weights=None):
     r"""
-    Compute the first Wasserstein distance between two 1D distributions.
+    Compute the general wasserstein-1 function between two n dimensional
+    ditributions(histograms).
```
okay will do that.

com3dian committed:
Revert "ENH-17473"

This reverts commit eca3095afbfcee68dc6bb2b7bd3371b9a5f67a1a.

com3dian committed:
ENH-17473

* clean the comments of stats.wasserstein_distance function

* fix rendering issue of stats.wasserstein_distance doc

com3dian committed:
Merge branch 'main' of https://github.com/scipy/scipy into feature_emd_gh17290

com3dian committed:
ENH-17473

* fix lint error(line 9816)

com3dian committed:
ENH-17473

* clean up comments of stats.wasserstein_distance function

com3dian committed:
Merge branch 'main' of https://github.com/scipy/scipy into feature_emd_gh17290

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9769,9 +9835,58 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
+    u_values, u_weights = _validate_distribution(u_values, u_weights)
+    v_values, v_weights = _validate_distribution(v_values, v_weights)
+
+    # convert value into 2D array
+    m, n = len(u_values), len(v_values)
+    u_values = asarray(u_values).reshape(m, -1)
+    v_values = asarray(v_values).reshape(n, -1)
+
+    # if dimensions are not equal throw error
+    if u_values.shape[1] != v_values.shape[1]:
+        raise ValueError('Dimensions of value are not equal.')
+
+    # if data is 1D then call the cdf_distance function
+    if u_values.shape[1] == 1 and v_values.shape[1] == 1:
+        return _cdf_distance(1,
+                             u_values.flatten(), v_values.flatten(),
+                             u_weights, v_weights)
```
This all looks like validation (and conversion to a standard form). Please move it to `_validate_distribution`.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9769,9 +9835,58 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
     >>> wasserstein_distance([3.4, 3.9, 7.5, 7.8], [4.5, 1.4],
     ...                      [1.4, 0.9, 3.1, 7.2], [3.2, 3.5])
     4.0781331438047861
-
+    >>> wasserstein_distance([[0, 2, 3], [1, 2, 5]], [[3, 2, 3], [4, 2, 5]])
+    3.0
+    >>> wasserstein_distance([[0, 2.75], [2, 209.3], [0, 0]],
+    ...                      [[0.2, 0.322], [4.5, 25.1808]],
+    ...                      [0.4, 5.2, 0.114], [0.8, 1.5])
+    174.15840245217169
     """
-    return _cdf_distance(1, u_values, v_values, u_weights, v_weights)
+    u_values, u_weights = _validate_distribution(u_values, u_weights)
+    v_values, v_weights = _validate_distribution(v_values, v_weights)
```
To avoid repeating `_validate_distribution` work, consider checking whether the inputs are 1d at the very top. If so, return `_cdf_distance`. This will make it more obvious that there are no changes to the existing behavior of the function.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9769,9 +9835,58 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
+    # if data is 1D then call the cdf_distance function
+    if u_values.shape[1] == 1 and v_values.shape[1] == 1:
+        return _cdf_distance(1,
+                             u_values.flatten(), v_values.flatten(),
```
```suggestion
                             u_values.ravel(), v_values.ravel(),
```
If you don't need to copy the data, use `ravel`. ([`flatten`](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.flatten.html) copies; [`ravel`](https://numpy.org/doc/stable/reference/generated/numpy.ravel.html) tries to reshape.)

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9769,9 +9835,58 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
+    elif np.any(np.isinf(u_values)) and np.any(np.isinf(v_values)):
+        return np.nan
+
+    # creat constraints
```
```suggestion
    # create constraints
```

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -7209,10 +7223,20 @@ def test_combine_weights(self):
                 [1, 1, 1, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1, 1]),
             stats.wasserstein_distance([5, 0, 1], [0, 4, 3],
                                        [1, 2, 4], [1, 2, 4]))
```
If the test above were under review today, I would have suggested:
```python3
rng = np.random.default_rng(4437304399483663605)
n_values, max_repeats = 10, 10
repeats = rng.integers(0, max_repeats, size=n_values)
u_values = rng.random(size=n_values)
u_weights = repeats * rng.random()  # overall scale of weights doesn't matter
v_values = rng.permuted(np.repeat(u_values, repeats))
v_weights = np.ones_like(v_values) * rng.random()
res = stats.wasserstein_distance(u_values, v_values, u_weights, v_weights)
assert_allclose(res, 0, atol=1e-15)
assert np.any(repeats == 0)  # true even for an edge cases like this
```

This is a property-based test that emphasizes the relationships that must be true without the distraction of the specific numerical values involved. It's easier for me to review because I only have to review the code rather than dozens of individual numbers. It ensures that the values weren't hand-selected to make the test pass, and the tests written like this tend to be stronger because it's easy to scale the size of the test and the number of instances tested. Note that I also tested that the total weight doesn't matter, fractional weights work as we'd expect, and I checked an edge case (one of the weights is zero, which is unusual).

`test_shift_nD` and `test_rotation_reflection_nD` already seem to be written like this. Great! Would you mind converting the other tests to property-based tests, since they are all checking relationships, not something to do with the particular numbers involved? If there are cases that don't lend themselves well to this style of test, or if the code would be much more complex than just hard-coding values, let me know, but I'm guessing that most will be amenable.

Another property-based test I'd like to see, if it is not already here (just point it out if it is): applying `wasserstein_distance` to 2D column arrays (and thus calculating the result with `milp`) gives the same distance as applying it to the flattened arrays (so that the distance is computed with `_cdf_distance`.

com3dian replied:
Thanks. I can get most of your points but can we clarify a bit about the code you provided? 
The last line
``` python
assert np.any(repeats == 0)  # true even for an edge cases like this
```
raise an Assertion error if the `repeats` variable has no zero values, right? Does this has any relationship with the `wasserstein` function behavior?

>Another property-based test I'd like to see, if it is not already here (just point it out if it is): applying wasserstein_distance to 2D column arrays (and thus calculating the result with milp) gives the same distance as applying it to the flattened arrays (so that the distance is computed with _cdf_distance.

I think `test_multi_dim` is what you would like to have.

mdhaber replied:
> raise an Assertion error if the repeats variable has no zero values, right? I don't see how the wasserstein function behavior can affect this.

Correct in both cases. Another way to do that is to manually add a zero to `repeats`. (Would that be better practice? I'm not sure.) Instead, since a zero is already there with the random seed I happened to generate, I chose to just highlight that they are already there.

You're right that `wasserstein_distance` can't affect the pass or failure of that line. The line is simplify verifying a feature of the test itself. If it fails due to a change of the test, then we are alerted that the test no longer covers this edge case.

com3dian replied:
Okay I see.
> since a zero is already there with the random seed I happened to generate, I chose to just highlight that they are already there.

but can you still call this a 'property based test' when it actually depends on your specific random seed :rofl: ?

mdhaber replied:
I'm not sure if that was rhetorical, so I'll answer. The test of `wasserstein_distance` itself is property-based. Precisely what the test checks depends on the seed and structure of the test, which in some sense could be said of any property-based test. If you wanted to add a stochastic but non-zero number of zeros to `repeats` at random positions, you could do that to make this feature of the test less dependent on the seed. This doesn't seem necessary to me because for most seeds there will happen to be at least one zero already. Also, the presence of a zero is not an essential element of the test - just something extra I noticed that the test already covered by the existing seed, which I generated at random. I hope you agree that either way there is a big improvement over what is there now.

com3dian replied:
I agree that either way can largely improve the current test. Yet I'm still questioning it because I'm not convinced that the zero values in a vector should be considered as a corner case. First the zero value is not special to `test_combine_weights` as they are very normal inputs like other positive and negative inputs (negative inputs are legal as long as their summed weight on a single value is positive and finite), and we already have another test `test_zero_weight` which is designed specifically for the zero weight case, so I would suggest we remove the last line of none zero vector assertion for simplicity and comprehensibility.

com3dian replied:
Also, I believe your proposed changes should go to `test_same_distribution`.

com3dian replied:
Yeah agree.

mdhaber reviewed (changes requested):
I haven't done a full review of the algorithmic part, but took a look at the tests. Here are a few other tests I'd need to see.
1) Input validation. Does the intended thing happen (e.g. error get raised) when you pass invalid input? (e.g. use `pytest.raises` with `match="specific error message I'm looking for"`.)
2) Edge cases. (For instance, what happens if one of the weights is zero? https://github.com/scipy/scipy/pull/17473/files#r1142812628)
3) Comparison against reference values for particular problems. (This is where the numbers used actually matter!) Are there published values of Wasserstein distances you can compare against? If not, can you compare against the output of POT?

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9769,9 +9835,58 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
+    # solving LP
+    constraints = LinearConstraint(A=A.T, ub=cost)
+    opt_res = milp(c=-b, constraints=constraints, bounds=(-np.inf, np.inf))
+    return b @ opt_res.x
```
```suggestion
    return -opt_res.fun
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9769,9 +9835,58 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
+
+    # get cost matrix
+    D = distance_matrix(u_values, v_values, p=2)
+    cost = D.reshape(n * m)
```
```suggestion
    cost = D.ravel()
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9769,9 +9835,58 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
+    cost = D.reshape(n * m)
+
+    # create the minimization target
+    p_u = np.ones(m)/m if u_weights is None else u_weights/np.sum(u_weights)
+    p_v = np.ones(n)/n if v_weights is None else v_weights/np.sum(v_weights)
```
```suggestion
    p_u = np.full(m, 1/m) if u_weights is None else u_weights/np.sum(u_weights)
    p_v = np.full(m, 1/n) if v_weights is None else v_weights/np.sum(v_weights)
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9769,9 +9835,58 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
+        return np.nan
+
+    # creat constraints
+    _row_func = np.vectorize(lambda x: np.floor(x/n) if x < m*n else m + x % n)
+    row = _row_func(np.arange(m * n * 2))
+    col = np.append(np.arange(m * n), np.arange(m * n))
+    value = np.ones(m * n * 2, dtype=float)
+    # sparse constraint matrix of size (m + n)*(m * n)
+    A = coo_matrix((value, (row, col)), shape=(n + m, m * n))
```
For readability, could you refactor this to use [sparse array creation functions](https://docs.scipy.org/doc/scipy/reference/sparse.html#functions)? For instance, the first part of the matrix is like:
```python3
from scipy.sparse import block_diag
A = np.ones((1, 2))  # row of two ones
B = (A,)*5  # five of these
C = block_diag(B)  # arranged as a block diagonal
```

And the second part
```python3
A = sparse.eye(2)  # 2x2 identity
B = (A,)*5  # five of these
C = sparse.hstack(B)  # arranged horizontally
```

Then you can stack these parts vertically with `vstack`.

If it doesn't give you too much trouble, we should probably also use [`coo_array`](https://docs.scipy.org/doc/scipy/reference/sparse.html#sparse-array-classes) instead of `coo_matrix` to try to stay up to date.

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -7249,6 +7294,88 @@ def test_inf_values(self):
+        assert_almost_equal(expected, shift_result)
+
+    @pytest.mark.parametrize('input_dim', (3, 4, 5, 22))
+    @pytest.mark.parametrize('data_volumn', ((6, 3), (10, 10), (30, 8),
+                                             (6, 150), (20, 14)))
+    def test_rotation_reflection_nD(self, input_dim, data_volumn):
+        # Rotating or reflecting across across n-dimensional
+        # space does not change the outcome.
+        rng = np.random.default_rng(734826773)
+
+        uu = rng.random((data_volumn[0], input_dim))
+        vv = rng.random((data_volumn[1], input_dim))
+        expected = stats.wasserstein_distance(uu, vv)
+        diag_value = rng.choice([-1, 1], size=input_dim)
+        # random reflection matrix
+        reflection_matrix = np.diag(diag_value)
+        combinations = []
+        for i in range(input_dim - 1):
+            for j in range(i + 1, input_dim):
+                combinations.append((i, j))
+
+        def rotation(index):
+            # generate random nD rotation matrix
+            # rotation is not uniformly distributed on unit sphere!
+            index_i, index_j = index
+            rotation_matrix = np.identity(input_dim)
+
+            rotation_angle = np.pi * rng.uniform(0, 1)
+            rotation_matrix[index_i, index_i] = np.cos(rotation_angle)
+            rotation_matrix[index_i, index_j] = - np.sin(rotation_angle)
+            rotation_matrix[index_j, index_i] = np.sin(rotation_angle)
+            rotation_matrix[index_j, index_j] = np.cos(rotation_angle)
+            return rotation_matrix
+        rotations = [rotation(index) for index in combinations]
+        rotation_matrix = np.linalg.multi_dot(rotations)
+        rotation_reflection = reflection_matrix @ rotation_matrix
+        uu_rr, vv_rr = uu @ rotation_reflection, vv @ rotation_reflection
+
+        shift_result = stats.wasserstein_distance(uu_rr, vv_rr)
+        assert_almost_equal(expected, shift_result)
```
These are great, but let's combine them. I think the random reflection/rotation matrix can be created by drawing a matrix from the orthogonal group.

```python3
rng = np.random.default_rng(3178853690258791578)
nu, nv, ndim = 10, 15, 5  # you can parameterize this part

u_values = rng.random(size=(nu, ndim))
v_values = rng.random(size=(nv, ndim))
u_weights = rng.random(size=nu)
v_weights = rng.random(size=nv)
ref = wasserstein_distance(u_values, v_values, u_weights, v_weights)

dist = stats.ortho_group(ndim)
transform = dist.rvs(random_state=rng)
shift = rng.random(size=ndim)
res = wasserstein_distance(u_values @ transform + shift,
                           v_values @ transform + shift,
                           u_weights, v_weights)
assert_allclose(res, ref)
```

mdhaber committed:
Merge branch 'main' into feature_emd_gh17290

mdhaber commented:
@com3dian There was a merge conflict so I resolved it for you. Are you interested in finishing this up?

com3dian commented:
Thanks a lot! I am currently actively working at my local branches, but there is still a lot of work to be done on the tests and documents, so it might take a few days before I can push again.

com3dian committed:
Merge branch 'scipy:main' into feature_emd_gh17290

com3dian committed:
ENH-17473

* complement wasserstein_distance doc

* cleaned up stats.wasserstein_distance function

* add property-base tests to TestWassersteinDistance

com3dian committed:
Revert "ENH-17473"

This reverts commit 25dfbdd1188c4e53600ab9fdbc242807398857bd.

com3dian committed:
ENH-17473

    * complement wasserstein_distance doc

    * cleaned up stats.wasserstein_distance function

    * add property-base tests to TestWassersteinDistance

com3dian committed:
managed imports

com3dian committed:
fixed 32bit numpy error

com3dian committed:
Merge branch 'scipy:main' into feature_emd_gh17290

mdhaber reviewed (commented):
There are some unresolved old comments:
- [x] https://github.com/scipy/scipy/pull/17473#discussion_r1142561691
- [x] https://github.com/scipy/scipy/pull/17473#discussion_r1142497160
- [x] https://github.com/scipy/scipy/pull/17473#pullrequestreview-1335329845

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -7361,6 +7415,72 @@ def test_inf_values(self):
+        add_dim = rng.integers(0, ndim, size = 1)[0]
+        add_value = rng.random()
+
+        u_values = np.c_[u_values[:, :add_dim], 
+                         np.full(nu, add_value), 
+                         u_values[:, add_dim:]]
+        v_values = np.c_[v_values[:, :add_dim], 
+                         np.full(nv, add_value), 
+                         v_values[:, add_dim:]]
```
```suggestion
        u_values = np.insert(u_values, add_dim, add_value, axis=1)
        v_values = np.insert(v_values, add_dim, add_value, axis=1)
```
Please use this idea in `test_zero_weight_nD`, too.

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -7336,6 +7357,20 @@ def test_collapse(self):
+        # test collapse for n dimentional values
+        # Collapsing a n-D distribution to a point distribution at zero
+        # is equivalent to taking the average of the norm of data.
+        u = np.arange(-10, 30, 0.4).reshape((-1, 10))
```
Is there something special about `arange` or can these be random?

com3dian replied:
No I think I was simply following the existing example there, already updated that with random values.

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -7336,6 +7357,20 @@ def test_collapse(self):
+            np.mean(np.linalg.norm(u, axis=1)))
+
+        u_weights = np.arange(len(u))
+        v_weights = u_weights[::-1]
```
Is there something special about these weights being reversed relative to the originals?

com3dian replied:
Also no, I have this one updated as well.

mdhaber reviewed (commented):
Tests are looking much better! Re: https://github.com/scipy/scipy/pull/17473#pullrequestreview-1349590254, thanks for generalizing all the property-based tests and adding (1) and (2). For testing, we need just one more thing - (3) - a comparison against reference values for particular problems. Actually, since the other tests are quite strong, I'd be happy with just one comparison against an outside reference.

com3dian committed:
Merge branch 'scipy:main' into feature_emd_gh17290

com3dian commented on scipy/stats/tests/test_stats.py:
```diff
@@ -7291,17 +7291,37 @@ def test_simple(self):
+        v_values = np.repeat(u_values, repeats, axis=0)
+        v_weights = rng.random(np.sum(repeats))
+        range_repeat = np.repeat(np.arange(len(repeats)), repeats)
+        u_weights = np.bincount(range_repeat, weights = v_weights)
```
Sure! i will check the linting throughout.

com3dian commented on scipy/stats/tests/test_stats.py:
```diff
@@ -7361,6 +7415,72 @@ def test_inf_values(self):
+            u_values = rng.random(size = (nu, 10))
+            v_values = rng.random(size = (nv, 2))
+            res = stats.wasserstein_distance(u_values, v_values)
+        excinfo.match('Invalid input values. Please provide two numpy'
```
Okay, already changed this.

com3dian committed:
Updated tests in TestWassersteinDistance

com3dian committed:
Merge branch 'feature_emd_gh17290' of github.com:com3dian/scipy into feature_emd_gh17290

mdhaber commented:
Thanks @com3dian. I'll provide the next round of review when you address:
- [ ] https://github.com/scipy/scipy/pull/17473#discussion_r1142561691
- [ ] https://github.com/scipy/scipy/pull/17473#discussion_r1142497160
- [ ] https://github.com/scipy/scipy/pull/17473#pullrequestreview-1335329845
- [ ] Add one (rather general) comparison against an external reference
- [ ] Replace the `np.concatenate` in `test_zero_weight_nD` with a `np.insert` (if that makes sense)

I don't expect to need to much more after that.

com3dian committed:
ENH-17473

* added explanation of inputs in the wasserstein_distance doc

* replaced np.concanate with np.insert in TestWassersteinDistance.test_zero_weight_nD

* replaced invalid value ingoration in TestWassersteinDistance.test_inf_values

com3dian committed:
Merge branch 'scipy:main' into feature_emd_gh17290

com3dian committed:
Merge branch 'feature_emd_gh17290' of github.com:com3dian/scipy into feature_emd_gh17290

com3dian committed:
ENH-17473

* removed empty lines in wasserstein_distance doc

* added value-based test in TestWassersteinDistance.test_pulished_values

com3dian commented:
Hi Matt @mdhaber ! Thank you for pointing those out! Below you can find a list of my commits that address the issues you raised in your reviews:

### :link:  https://github.com/scipy/scipy/pull/17473#discussion_r1142561691 
I have included an additional paragraph for clarification under the parameters explanation for `u_values`, `v_values`.


### :link:  https://github.com/scipy/scipy/pull/17473#discussion_r1142497160
Replaced invalid warning codes.


### :link:  https://github.com/scipy/scipy/pull/17473#discussion_r1142507769
I have included a new paragraph within the note section of the documentation for wasserstein_distance.


### :raised_hands:  Add one (rather general) comparison against an external reference
I firstly went to [POT](https://pythonot.github.io/) in Python and [transport](https://cran.r-project.org/web/packages/transport/index.html) in R, but neither of them has a solid multivariate example for the Wasserstein distance. I also checked if there is one for Julia and found [this](https://github.com/JuliaOptimalTransport/OptimalTransport.jl) but they seem don't have written examples at all. Then I asked the chatgpt empowered bing and luckily found [this one](https://jamesmccaffrey.wordpress.com/2018/03/05/earth-mover-distance-wasserstein-metric-example-calculation/). Although it's not a book or a paper, the blog is solid enough as it clearly showed the process of solving a optimal transport plan decomposed into steps. To be honest I'm afraid this one is probably the only non-trivial nD example in the  public internet, 

\*Note: the result in the blog is not precise, because the author keeps only two decimal places for the distance. To obtain more accurate results. To obtain more accurate results, I recalculated by using the precise values in the `test_pulished_values` method. The difference in distance is relatively small, so the accurate results from `wasserstein_distance` and the manually computed method (both == 2.4709551919132666) are comparable and close to the answer presented in original post (2.47) .

### :man_technologist:  Replace the np.concatenate in test_zero_weight_nD with a np.insert (if that makes sense)
Thanks a lot and updated.

-------------------------------------
:crossed_flags: BTW I found there are multiple errors in many different place both locally and on Azure, is that my problem? The local errors are all Cython errors, for instance:
```
ERROR scipy/stats/_page_trend_test.py - TypeError: C function scipy.linalg.cython_lapack.cgetrf has wrong signature...
```
I tried with updating/recreating my conda environment for Scipy development, and also ran `git submodule update --init`, yet my efforts did not fix the bugs.

mdhaber commented:
I don't think the errors are your fault. CI is broken for all PRs right now. I don't know about locally, though. If you are using Meson, I can't help. If you are using `setup.py` like me, try `git clean dxf` before building again.
Note that a published example is best, but but required. It's ok to just solve a random (but non-trivial) example with other software.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9703,14 +9710,14 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
                 v\\
             \end{bmatrix}
-
+    The :math:`\text{vec}()` function denotes the Vectorization fucntion.
```
```suggestion
    The :math:`\text{vec}()` function denotes the Vectorization function.
```
Although I've been doing this for a while, and I haven't heard of something referred to as a "Vectorization function".

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9703,14 +9710,14 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
     probability mass transported from :math:`u` to :math:`v`. Summing over
     the rows of :math:`\Gamma` should give the source distribution :math:`u` :
-    :math:`\sum_j \gamma_{ij} = u(x_i)` holds for all :math:`i`s and summing
+    :math:`\sum_j \gamma_{ij} = u(x_i)` holds for all :math:`i` s and summing
```
It's correct without the"s", and it will render better.
```suggestion
    :math:`\sum_j \gamma_{ij} = u(x_i)` holds for all :math:`i` and summing
```

com3dian replied:
I looked for references whether mathematical variable are plural or not but there seems are supports from both sides ([mathoverflow](https://mathoverflow.net/questions/192173/how-to-refer-to-plural-of-mathematical-symbols-with-or-without-an-apostrophe), [stackexchange](https://english.stackexchange.com/questions/337102/plural-for-math-notations), [stackexchange](https://ell.stackexchange.com/questions/167898/plural-of-a-mathematical-variable)). I personally feel like using plural forms of subscript variables is more natural, but I'm not native English speaking so I let you decide this one.

mdhaber replied:
> but I'm not native English speaking so I let you decide this one.

Thanks. "For all $x$" is a very common mathematical phrase, and "s" after the variable is uncommon in published work. I agree that with usual nouns we would need the "s" ("for all ducks...") but in this context I believe it is understood that we are talking about the _values_ that the variable can assume, e.g. "for all values of the variable $i$". Here are the first three academic texts about Wasserstein distance I found with the phrase "for all" (not selected to make a point), and "s" does not appear after the variable.

https://www.annualreviews.org/doi/full/10.1146/annurev-statistics-030718-104938
https://link.springer.com/article/10.1007/s13373-017-0101-1
https://library.oapen.org/bitstream/handle/20.500.12657/37704/1/2020_Book_AnInvitationToStatisticsInWass.pdf

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9703,14 +9710,14 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
+    :math:`\sum_j \gamma_{ij} = u(x_i)` holds for all :math:`i` s and summing
     over the columns of :math:`\Gamma` should give the target distribution
-    :math:`v`: :math:`\sum_i \gamma_{ij} = v(y_j)` holds for all :math:`j`s.
+    :math:`v`: :math:`\sum_i \gamma_{ij} = v(y_j)` holds for all :math:`j` s.
```
```suggestion
    :math:`v`: :math:`\sum_i \gamma_{ij} = v(y_j)` holds for all :math:`j`.
```

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -7293,6 +7293,25 @@ def test_simple(self):
             [0, 1, 2], [1, 2, 3]),
             1)
+    
+    def test_pulished_values(self):
```
```suggestion
    def test_published_values(self):
```

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -7293,6 +7293,25 @@ def test_simple(self):
             1)
+    
+    def test_pulished_values(self):
+        # Compare against pulished values and manually computed results.
```
```suggestion
        # Compare against published values and manually computed results.
```

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -7293,6 +7293,25 @@ def test_simple(self):
+    
+    def test_pulished_values(self):
+        # Compare against pulished values and manually computed results.
+        # The values and computed result are posted at James D. McCaffrey's bolg,
```
```suggestion
        # The values and computed result are posted at James D. McCaffrey's blog,
```

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -7293,6 +7293,25 @@ def test_simple(self):
+        
+        res = stats.wasserstein_distance(u, v)
+        # In original post, the author kept two decimal places for ease of calculation.
+        # This test uses the more precise value of distance to get the precise results,
+        # for comparison please look at the table and figure in the priginal blog.
```
```suggestion
        # This test uses the more precise value of distance to get the precise results.
        # For comparison, please see the table and figure in the original blog post.
```

mdhaber reviewed (commented):
Thanks. I'll try to complete my review tomorrow.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9621,19 +9626,34 @@ def combine_pvalues(pvalues, method='fisher', weights=None):
 
     Parameters
     ----------
     u_values, v_values : array_like
```
How about:
```
    u_values : 1d or 2d array_like
        A sample from a probability distribution or the support (set of all
        possible values) of a probability distribution. Each element along
        axis 0 is an observation or possible value. If two-dimensional, axis
        1 represents the dimensionality of the distribution; i.e., each row is
        a vector observation or possible value.
    v_values : 1d or 2d array_like
        A sample from or the support of a second distribution.
    u_weights, v_weights : 1d array_like, optional
        Weights or counts corresponding with the sample or probability masses
        corresponding with the support values. Sum of elements must be positive
        and finite.
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -37,7 +37,12 @@
 
 from scipy.spatial.distance import cdist
+from scipy.spatial import distance_matrix
+from scipy.sparse import hstack, vstack, coo_array, block_diag, eye
```
```suggestion
from scipy.sparse import coo_array, block_diag
```
It's OK with me to import these, but in the code below, it would not be immediately obvious whether `hstack`, etc. are for NumPy arrays or sparse arrays. Personally, I'd suggest:
```
from scipy import sparse
```
and always using fully-qualified names.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9621,19 +9626,34 @@ def combine_pvalues(pvalues, method='fisher', weights=None):
-    that must be moved, multiplied by the distance it has to be moved.
+    The Wasserstein distance, also called the Earth mover's distance or the
+    optimal transport distance, is defined as a similarity metric between
+    two probability distribution. In the discrete case, the Wasserstein
```
```suggestion
    two probability distributions. In the discrete case, the Wasserstein
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9649,8 +9669,9 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
     -----
-    The first Wasserstein distance between the distributions :math:`u` and
-    :math:`v` is:
+    Given two samples (expressed as probability mass functions), :math:`u`
+    and :math:`v`, the first Wasserstein distance between the distributions
```
To simplify
```suggestion
    Given two probability mass functions, :math:`u`
    and :math:`v`, the first Wasserstein distance between the distributions
```
It's not immediately clear what it means to express a sample as a probability mass function. If you want to also define the Wasserstein distance means for two samples, I would do that after defining what it means for distributions.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9677,9 +9761,19 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
+    .. [4] Peyré, Gabriel, and Marco Cuturi. "Computational optimal
+           transport." Center for Research in Economics and Statistics
+           Working Papers 2017-86 (2017).
+    .. [5] Vincent Herrmann's blog
+           https://vincentherrmann.github.io/blog/wasserstein/
```
```suggestion
    .. [5] Hermann, Vincent. "Wasserstein GAN and the Kantorovich-Rubinstein
           Duality". https://vincentherrmann.github.io/blog/wasserstein/.
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9691,9 +9785,59 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
+    3.0
+    >>> wasserstein_distance([[0, 2.75], [2, 209.3], [0, 0]],
+    ...                      [[0.2, 0.322], [4.5, 25.1808]],
+    ...                      [0.4, 5.2, 0.114], [0.8, 1.5])
```
```suggestion
    ...                       [0.4, 5.2, 0.114], [0.8, 1.5])
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9691,9 +9785,59 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
     ...                      [1.4, 0.9, 3.1, 7.2], [3.2, 3.5])
     4.0781331438047861
-
+    >>> wasserstein_distance([[0, 2, 3], [1, 2, 5]], [[3, 2, 3], [4, 2, 5]])
```
We don't have to improve the old documentation, but it's not a bad idea to describe what the code is doing.
```suggestion
Compute the Wasserstein distance between two three-dimensional samples,
each with two observations.

    >>> wasserstein_distance([[0, 2, 3], [1, 2, 5]], [[3, 2, 3], [4, 2, 5]])
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9691,9 +9785,59 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
-
+    >>> wasserstein_distance([[0, 2, 3], [1, 2, 5]], [[3, 2, 3], [4, 2, 5]])
+    3.0
+    >>> wasserstein_distance([[0, 2.75], [2, 209.3], [0, 0]],
```
```suggestion
Compute the Wasserstein distance between two two-dimensional distributions
with three and two (weighted) observations, respectively.

    >>> wasserstein_distance([[0, 2.75], [2, 209.3], [0, 0]],
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9691,9 +9785,59 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
+    174.15840245217169
     """
-    return _cdf_distance(1, u_values, v_values, u_weights, v_weights)
+    m, n = len(u_values), len(v_values)
+    u_values = asarray(u_values).reshape(m, -1)
+    v_values = asarray(v_values).reshape(n, -1)
```
I'd prefer that you just convert to arrays here, then next you just check whether both `ndim` are 1.

(After that, you can do all the rest of this stuff - getting `m`, `n` and ensuring that both arrays are 2d. I'd suggest raising an error if they have more than two dimensions.)

mdhaber replied:
> because the user could pass 2d arrays that actually could be computed with the 1d _cdf_distance , and which would be much faster

Yes, I know. I don't think it matters much either way.
* Option A (yours): faster in the unusual case that a user passes an `(n, 1)` array
* Option B (mine): *slightly* simpler (`ndim` instead of `shape[-1]` and no `ravel()`) and allows for the possibility of a unit test that compares optimization against `_cdf_distance`.

I considered both options before making the suggestion and thought that the benefits of B outweighed the benefits of option A. If you'd like to wait for someone else's opinion, we can do that, but I'd suggest that this is a minor point that should not hold up the PR. You can always submit a PR with optimizations like this after the initial PR merges; it would be a backward compatible change.

com3dian replied:
Agreed:) I will keep it option B in this PR then.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9691,9 +9785,59 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
+    v_values = asarray(v_values).reshape(n, -1)
+
+    # if data is 1D then call the cdf_distance function
+    if u_values.shape[1] == 1 and v_values.shape[1] == 1:
+        return _cdf_distance(1,
+                             u_values.ravel(), v_values.ravel(),
+                             u_weights, v_weights)
```
```suggestion
    if u_values.ndim == 1 and v_values.ndim == 1:
        return _cdf_distance(1, u_values, v_values, u_weights, v_weights)
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9691,9 +9785,59 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
+
+    # if dimensions are not equal throw error
+    if u_values.shape[1] != v_values.shape[1]:
+        raise ValueError('Invalid input values. Please provide two numpy'
+                         'arrays with either both one-dimensional or both'
+                         'two-dimensional arrays with the same number of'
+                         'columns.')
```
At this point, we know that the user was not trying to pass one-dimensional arrays. Also, it's ambiguous whether the arrays must have the same number of columns if they're one-dimensional.
```suggestion
        raise ValueError('Invalid input values. If two-dimensional,'
                         '`u_values` and `v_values` must have the same'
                         'number of'columns.')
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9691,9 +9785,59 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
+                         'two-dimensional arrays with the same number of'
+                         'columns.')
+
+    # if data contains np.inf then return inf or nan
```
How about weights?

com3dian replied:
`_validate_distribution` takes care of the nan or inf values I think, it will raise a value error.

https://github.com/scipy/scipy/blob/39ef81242412f307a4c5c7ac0170b81465954388/scipy/stats/_stats_py.py#L10016-L10019

mdhaber reviewed (commented):
Looks good @com3dian. Just a few more thoughts from me.

@steppi or @tirthasheshpatel I'm reviewing the rest, but would one of you be willing to double-check the math documentation and main implementation (after input validation is done) here?

com3dian committed:
Merge branch 'scipy:main' into feature_emd_gh17290

com3dian committed:
Update scipy/stats/_stats_py.py

Co-authored-by: Matt Haberland <mhaberla@calpoly.edu>

com3dian committed:
ENH-17473

* add error code tests in TestWassersteinDistance.test_error_code

* fixed typo in/updated wasserstein_distance doc

* removed inputs reshapping and added ndim condition statement in wasserstein-distance

com3dian committed:
Merge branch 'scipy:main' into feature_emd_gh17290

com3dian committed:
fixed imports

com3dian committed:
updated imports

com3dian committed:
fix ndim comparison in wasserstein_distance and np.insert error in TestWassersteinDistance.test_zero_weights

com3dian committed:
fixed render issue

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -9770,16 +9784,81 @@ def wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None):
+                u\\
+                v\\
+            \end{bmatrix}
+    The :math:`\text{vec}()` function denotes the Vectorization function 
```
```suggestion

    The :math:`\text{vec}()` function denotes the Vectorization function 
```

com3dian replied:
Thanks, I will have a look at this later today. Also, is it possible I could test the doc locally like rendering only the specific page?

mdhaber replied:
Try `python dev.py doc` to build docs locally. If you have trouble, follow the latest build instructions in this PR: https://github.com/scipy/scipy/pull/18429

com3dian committed:
fixed rendering issue

com3dian committed:
fixed rendering issue

mdhaber commented:
Looks like that worked. Please let me know when you've responded to all the comments.

com3dian commented:
Yeah I think I finished with this round @mdhaber . The updates includes
- dimension checks in the `wassertsein_distance` function
- documentation updates (in example and in notes)
- tests updated with new error codes for dimension difference

com3dian committed:
removed duplicatd lines in doc

mdhaber commented:
But there are still four unaddressed comments.

com3dian committed:
updated doc in wasserstein_distance function

com3dian committed:
removed extra empty line

com3dian committed:
Updated wasserstein_ditance fucntion doc

steppi commented:
I should be able to get to this sometime this week.

com3dian commented:
Sure @mdhaber , just want to make sure the current version is all good and thanks both for answering @steppi !

mdhaber commented:
Ping @tirthasheshpatel

mdhaber reviewed (approved):
All my comments from https://github.com/scipy/scipy/pull/17473#pullrequestreview-1407356972 and before were addressed, and @tirthasheshpatel agreed with the math, so this looks good to me. 

I asked @tirthasheshpatel to wait to merge, though, because I wanted to reconsider https://github.com/scipy/scipy/issues/17290#issuecomment-1310557827. I think this would be the first reducing function that accepts a multivariate sample. Right now it just accepts a single sample, but many other stats functions accept Nd arrays that represent several independent univariate samples, with an `axis` argument to specify the axis along which reduction is to be performed. Without a convention, this is problematic because there is ambiguity over whether a 2D array represents a 1D collection of multivariate observations or a 2D array of univariate observations. There are a few ways we could go here, e.g.:

1. Not vectorize multivariate-sample reducing functions
2. Stick with the convention we've implicitly adopted for multivariate elementwise functions/methods, like `mutlivariate_normal.pdf`, in which the last axis corresponds with the dimensionality. In this case `axis` could be any axis other than the last one.
3. Accept two conceptually distinct axis arguments

I have run into this problem before - gh-17781 - so I think we need to at least talk about this before merging. 

_Update: there is some precedent set by [`directional_stats`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.directional_stats.html)._

tirthasheshpatel reviewed (approved):
Confirmed that the math and the implementation here matches with the blog. The tests look very comprehensive and strong. Docs also look good.

About adding the `axis` argument: I think it's not worth all the work to add + maintain it where we allow multi-variate inputs. Option 2 sounds like the way to go if we add vectorization support in the future.

com3dian commented:
You are welcome @mdhaber :smile:  !

fuglede commented:
I'm way too late for the party here, but I wonder if this could've also been treated as a special case of linear assignment, so that the more specialized (and generally much faster) `scipy.optimize.linear_sum_assignment`, or, in the sparse case, `scipy.sparse.csgraph.min_weight_bipartite_full_matching`, could have also done the job? See e.g. https://stackoverflow.com/a/57563383
